sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs

```diff
@@ -558,25 +558,49 @@ private Block BuildTryStatement(TryStatementSyntax tryStatement, Block currentBl
                 tryEndStatementConnections.Add(this.ExitTarget.Peek());
             }
 
-            var tryBody = BuildBlock(tryStatement.Block,
-                CreateBranchBlock(tryStatement, tryEndStatementConnections.Distinct()));
-
-            var tryStartStatementConnections = catchBlocks.ToList();
-            tryStartStatementConnections.Add(tryBody); // try body
-
-            if (!areAllExceptionsCaught) // unexpected exception thrown, go to exit (through finally if present)
+            Block tryBody;
+            if (tryStatement.Block.Statements.Any(s=>s.IsKind(SyntaxKind.ReturnStatement)))

```

Comment:
before this change, when there was a `return` statement inside the body of the `try`, it would short-circuit the `catch` blocks, having only one possible successor: the `exit` block. This was wrong, because if an exception is thrown before the `return`, it can still reach `catch` blocks

Reply:
Will `tryStatement.Block.Statements.Any(...)` find all `return` statements inside the `try-catch`, or only those that are direct children?

Reply:
Good point, I'll add a test for this use case